### DISCUSSION TITLE
MOBILE - How to press a next button and focus another text field input?

### DISCUSSION THREAD

**Discussion by vagnerwentz** [react-hook-form/react-hook-form#8059] [Q&A]
### Version react hook form
* ^7.27.0

### What I tried to follow and without successful
[react hook form - Discussions 7818](https://github.com/react-hook-form/react-hook-form/discussions/7818)
[react hook form - Issues 230](https://github.com/react-hook-form/react-hook-form/issues/230)

### About what I have
I have 4 text field components at my screen, the name of each text field is `name, documentation, email, password` and I would like to know how can I setup some configuration that it will be pressed the NEXT button at keyboard and will focus the following text fields?

An example that I have inside at my component file, again I would like to press the next button and the next component that I will config, will be focus.
```tsx
<TextField
  name="name"
  label={I18n.t('registerPersonal.fullNameLabel')}
  placeholder={I18n.t('registerPersonal.fullNameInput')}
  icon={<TypographyIcon fill={!!errors.name && theme.colors.attention} />}
  error={errors.name?.message}
  errors={errors}
  control={control}
  returnKeyType="next"
/>

<TextField
  name="documentation"
  label={I18n.t('registerPersonal.documentIdentificationLabel')}
  placeholder={I18n.t('registerPersonal.documentIdentificationInput')}
  icon={
    <DocumentIcon
      fill={!!errors.documentation && theme.colors.attention}
    />
  }
  error={errors.documentation?.message}
  control={control}
  returnKeyType="next"
/>
```

Some properties that I get at my personal hook
```ts
const {
    control,
    handleSubmit,
    formState: { errors, isValid }
  } = useForm({ resolver: yupResolver(schema) })
```

My component TextField
```tsx
import { TextInputProps, Text } from 'react-native'
import { Control, useController } from 'react-hook-form'

import { Container, Wrapper, TextInput, Label } from './styles'

import theme from '../../global/styles/theme'

type TextFieldProps = {
  placeholder?: string
  label?: string
  icon?: React.ReactNode
  error?: string
  errors?: {
    [x: string]: any
  }
  name: string
  control: Control
} & TextInputProps

export function TextField(props: TextFieldProps) {
  const { placeholder, label, icon, error, errors, name, control, ...rest } =
    props

  const { field } = useController({
    control,
    defaultValue: '',
    name
  })

  return (
    <>
      <Container>
        {!!label && <Label>{label}</Label>}
        <Wrapper hasLabel={!!label} hasError={!!error}>
          {!!icon && icon}
          <TextInput
            value={field.value}
            onChangeText={field.onChange}
            placeholder={error ? error : placeholder}
            placeholderTextColor={
              error ? theme.colors.attention : theme.colors.grayColor
            }
            {...rest}
          />
        </Wrapper>
        {errors && errors.name && errors.name.type === 'matches' && (
          <Text>{errors.name.message}</Text>
        )}
      </Container>
    </>
  )
}

```

**Comment by Moshyfawn** [comment]
Have you tried [the `setFocus` API](https://react-hook-form.com/api/useform/setfocus)?
If you pass down the `field.ref` into the `TextInput` component, RHF will be able to manage the input focus and you can call `setFocus` in a, say, next button `onClick` callback.
```diff
  const { field } = useController({ name: "fieldName" })
  
  return (
    <TextInput
+    ref={field.ref}
    />
  )
```
```js
const { setFocus } = useForm()

onSubmitEditing={() => setFocus("fieldName")}
```

**Reply by vagnerwentz** [reply]
The `ref` property doesn't exist at my component, why? I don't know.

```
Type '{ name: string; ref: any; label: string; 
placeholder: string; icon: Element; error: any; errors: { [x: string]: any; }; 
control: Control<FieldValues, object>; returnKeyType: "next"; }' 
is not assignable to type 'IntrinsicAttributes 
& { placeholder?: string; label?: string; icon?: ReactNode; error?: string; errors?: { [x: string]: any; }; name: string; control: Control<FieldValues, object>; } & TextInputProps'.
  Property 'ref' does not exist on type 'IntrinsicAttributes & { placeholder?: string; label?: string; icon?: ReactNode; error?: string; errors?: { [x: string]: any; }; name: string; control: Control<FieldValues, object>; } & TextInputProps'.
```

**Reply by Moshyfawn** [reply]
Huh, you're passing a `ref` to the `TextInput` component from the `react-native` package?
See [this super simple Snack](https://snack.expo.dev/@moshyfawn/nb-rhf)

**Reply by Moshyfawn** [reply]
You might need to `React.forwardRef` with your custom `TextField` component down to the `TextInput`

**Reply by vagnerwentz** [reply]
My TextInput is a component that I created

**Reply by Moshyfawn** [reply]
Whatever the underling input component you're using to compose your `TextInput`, you want to get it's `ref` in order to manage the input's focus programmatically.

Maybe, share some part of your `TextInput` component implementation?

**Reply by vagnerwentz** [reply]
I put the TextInput implementation, is the last code at post

**Reply by Moshyfawn** [reply]
I see that the `TextInput` component comes from `./styles` but is it `import { TextInput } from "react-native"` in styles file or what's the implementation?
```jsx
import { TextInput } from './styles'
...
  return (
    ...
    <TextInput
      value={field.value}
      onChangeText={field.onChange}
      placeholder={error ? error : placeholder}
      placeholderTextColor={
        error ? theme.colors.attention : theme.colors.grayColor
      }
      {...rest}
    />
    ...
)
```

I'm able to pass down the `ref` callback to the `TextInput` from `react-native` in [this Snack](https://snack.expo.dev/@moshyfawn/nb-rhf). If you use some other component underneath, you might wanna check it's API for the `ref` prop _(whatever it's called; can be `inputRef` or smth else)_. If it's just the plain `<input />` element, it still takes in the `ref` prop

**Reply by vagnerwentz** [reply]
Yeah, the component TextInput at the styles, is a TextInput

**Reply by vagnerwentz** [reply]
@Moshyfawn Did you find something how can I do what I want?

**Reply by Moshyfawn** [reply]
Only that you should be able to get the `ref` of the `react-native` `TextInput` component via the `ref` component. IDK your `TextInput` implementation, but it's possible, seeing the error you've posted, that you just have to update the `TextInput`'s props types down in the styles file.

<img width="820" alt="image" src="https://user-images.githubusercontent.com/16290753/159223325-52a2ab02-993b-4fa7-a2f0-79f912361c6a.png">

You can't manage input's focus programmatically without having it's `ref`. And it's not the RHF limitation.

**Reply by Noitidart** [reply]
@Moshyfawn is it possible to put custom data into that ref? I did the below which works but I'm not sure if it's safe, seems like implementation detail.

I am also using React Native.

```
const controller = useController(...);
const inputRef = useRef();

controller.field.ref({
  focus: () => {
    // do my custom stuff here. like focus, and other stuff
  }
});

return <TextInput ref={inputRef} />;
```


I figured this out by `console.log`'ing `controller.field.ref.toString()` which showed it was:

```
         function (e) {
          var t = d(a._fields, s);
          e && t && e.focus && (t._f.ref = {
            focus: function focus() {
              return e.focus();
            },
            setCustomValidity: function setCustomValidity(t) {
              return e.setCustomValidity(t);
            },
            reportValidity: function reportValidity() {
              return e.reportValidity();
            }
          });
        }
```

**Reply by Moshyfawn** [reply]
I'm not actually sure, hm.

@bluebill1049, what's the reasoning for RHF behind [exposing only](https://github.com/react-hook-form/react-hook-form/blob/09d51cf70073efd296eac71981097efe47d83cf8/src/useController.ts#L130=) the `focus`, `setCustomValidity` and `reportValidity` properties in a `ref` from an error _(`fieldState.error?.ref`)_?

```ts
field._f.ref = {
  focus: () => elm.focus(),
  setCustomValidity: (message: string) =>
    elm.setCustomValidity(message),
  reportValidity: () => elm.reportValidity(),
};
```

**Reply by Noitidart** [reply]
Much love brother for the help.

**Reply by bluebill1049** [reply]
for consistent API with controlled component (Controller/useController)

**Reply by Noitidart** [reply]
In order to expose some custom data that I can use when showing which errors happened. I abused `setCustomValidity` like this, do you guys have any other recommendation? As if implementation changes it will break what I'm doing - https://github.com/react-hook-form/react-hook-form/discussions/8161#discussion-3999584
